### ISSUE TITLE
Data storage for learning

### ISSUE THREAD

**Issue by afonsocastro** [afonsocastro/larcc_interface#7]
Further required developments on [data_acquisition_node](https://github.com/afonsocastro/larcc_interface/blob/master/data_storage/src/data_aquisition_node.py):

- [x] Creating a proper time variable (seconds.nseconds), and assign the median of the 3 topics timestamps to our self timestamp
- [x] Initializing gripper communication and extract gripper information ("actual_force_motor_current" + "object_detected") (example on [force_torque_node](https://github.com/afonsocastro/larcc_interface/blob/master/gripper/src/force_torque_node.py))
- [x] Storing every Data For Learning in a proper file (SQLite, JSON, txt, ...)

The stored data should create its own file with a counter on its name: e.g., sample1, sample2, and so on. 
We will probably need to distinguish the training experiments with different poses. As so, I suggest saving names with this information: e.g., pos1_sample1, pos1_sample2, pos2_sample1, and so on.
Each different gesture should have its on folder: e.g., push, pull, twist.

**Comment by Joel-Baptista** [comment]
I was encountering some problems importing the RobotiqHand class from the gripper package to this script that is found in a different package. I solve it following the steps in this site (https://roboticsbackend.com/ros-import-python-module-from-another-package/). This procedure was used in the lib package as well to make it accessible to all scripts within the project

**Comment by Joel-Baptista** [comment]
I was not able to import the RobotiqHand class from the gripper package after all. In this version, the script is running with a copy of RobotiqHand and GripperStatusDTO while I can't fix this issue. However, this does not compromise the script's functionality.

Additionally, this script already has the behavior of starting storing the data in the beggining of the force's application and stopping saving data in the end of the force's application. The diagram explains the current scripts logic.

![Captura de ecrã de 2022-07-05 15-58-45](https://user-images.githubusercontent.com/58263121/177358127-98cbc570-c0e4-4ae9-b389-e9c998e9f184.png)

**Comment by afonsocastro** [comment]
This behavior seems to be appropriate for our intent! 
I want to try it :)

I am feeling that this method will save a different number of values for each different experiment, am I right?
Distinct applied forces for training will have distinct number of stored values... We should have this is mind for future developments.

Good work!

**Comment by afonsocastro** [comment]
Also, I have a new idea of how we can start importing modules from different folders with no problem at all:
We should mark the folder with the module of interest as a Sources Root.

In PyCharm Community we can simply:

1. right-click any folder in your project 
2. select "Mark Directory As"
3. select "Sources Root"

Modules within that folder will now be available for import. Any number of folders can be so marked.

@Joel-Baptista let's discuss this topic later. I feel the need of standardize this importing modules question.

**Comment by afonsocastro** [comment]
![image](https://user-images.githubusercontent.com/47828797/177856061-6e25e1a1-f1f0-435e-aa3a-08257ee5811d.png)

**Comment by Joel-Baptista** [comment]
Trying to figure out the problems with the timestamps I found something. There might be a problem with the timestamps that where given by the driver nodes. The image prints the max timestamp of the 3 nodes, but that's irrelevant since they are very close. Notice that when the number of seconds changes, it occurs same error, which is the next timestamp being larger than it should be. For example, the second timestamp is 1657276314.9804497, the last timestamp with the 4 digit, and the next timestamp is 1657276315.804361. 


![Captura de ecrã de 2022-07-08 11-32-48](https://user-images.githubusercontent.com/58263121/177976583-4526bf3b-6b3a-4409-81db-317c468b13cc.png)

**Comment by Joel-Baptista** [comment]
Some notes about the data storage for learning. The way the timestamps where measured caused problems. These problems are exposed in the comment before this one. In the effort of solving this problem, a new approach was taken. Now, the timestamp is measured when the message arrives from the topic to the data_aquisition_node, instead of reading the timestamp on the header of the message. This adds an error to the timestamp measure of about 1 to 3 milisseconds, which is a precision degree that can be ignored.

**Comment by Joel-Baptista** [comment]
I have some information about the normalize function used in the data which is from the sklearn. It is not ideal but it is not as bad. Analyzing by column, the function checks the number with the largest absolute value, and divides all number by it. This means that zeros remain zeros, negatives remain negatives and positives remain positves. This also means that the data doesn't go from -1 to 1 in every sample, however it is assured that there aren't any numbers larger than 1 and smaller than -1. See image for example

![Captura de ecrã de 2022-07-14 19-18-25](https://user-images.githubusercontent.com/58263121/179054299-af36d242-d9ee-4bbc-8e1b-4ee67af081ee.png)

**Comment by Joel-Baptista** [comment]
When analyzing the forces and torques in the gripper sensors, a relevant detail was discovered. The values of the sensors are not static when the robot is at rest, fluctuating in some sort of normal distribution. There were done various test in an attempt to quantify this lack of precision and understand its nature. The figure in this comment shows a frequency graph for all 6 values for one of the test. The tests took 10 seconds in a 100Hz frequency, to mimic the training scene. The red line represents the average of the values and the yellow lines represent the confidence interval of the average value at a 99.9%
![sensor_experiment](https://user-images.githubusercontent.com/58263121/181064604-7ae94a89-ad06-4fde-b9f5-8166f95cc9af.png)

**Comment by Joel-Baptista** [comment]
I'll leave this image here as confirmation that the values of the matrices correspond to the notes taken in the experimentation day 
![Splitted_data](https://user-images.githubusercontent.com/58263121/191029420-e2e68eb8-375e-44c7-8bd5-8c5dd83cd2f0.png)
